include/fuzzuf/algorithms/afl/templates/afl_mutation_hierarflow_routines.hpp

```diff
@@ -17,536 +17,549 @@
-    state.stage_max = stage_max_multiplier * perf_score / state.havoc_div / 100;
-    state.stage_cur_byte = -1;
-
-    if (state.stage_max < option::GetHavocMin(state)) {
-        state.stage_max = option::GetHavocMin(state);
-    }
+    s32 stage_max_multiplier,  // see directly below
+    int stage_idx) {
+  state.stage_name = stage_name;
+  state.stage_short = stage_short;
+  state.stage_max = stage_max_multiplier * perf_score / state.havoc_div / 100;
+  state.stage_cur_byte = -1;
 
-    u64 orig_hit_cnt = state.queued_paths + state.unique_crashes;
+  if (state.stage_max < option::GetHavocMin(state)) {
+    state.stage_max = option::GetHavocMin(state);
+  }
 
-    u64 havoc_queued = state.queued_paths;
+  u64 orig_hit_cnt = state.queued_paths + state.unique_crashes;
 
-    /* We essentially just do several thousand runs (depending on perf_score)
-       where we take the input file and make random stacked tweaks. */
+  u64 havoc_queued = state.queued_paths;
 
-    for (state.stage_cur = 0; state.stage_cur < state.stage_max; state.stage_cur++) {
-        using afl::util::UR;
+  std::array<u64, fuzzuf::mutator::NUM_CASE> havoc_operator_finds;
+  bool useHavocOperatorFinds =
+      fuzzuf::optimizer::Store::GetInstance().Exists(
+          fuzzuf::optimizer::keys::SelectedCaseHistogram) &&
+      fuzzuf::optimizer::Store::GetInstance().Exists(
+          fuzzuf::optimizer::keys::SelectedCaseHistogram);
 
-        u32 use_stacking = 1 << (1 + UR(option::GetHavocStackPow2(state), state.rand_fd));
+  /* We essentially just do several thousand runs (depending on perf_score)
+     where we take the input file and make random stacked tweaks. */
 
-        state.stage_cur_val = use_stacking;
-        mutator.Havoc(use_stacking, state.extras, state.a_extras, mutop_optimizer, custom_cases);
+  for (state.stage_cur = 0; state.stage_cur < state.stage_max;
+       state.stage_cur++) {
+    using afl::util::UR;
 
-        if (this->CallSuccessors(mutator.GetBuf(), mutator.GetLen())) return true;
+    u32 use_stacking =
+        1 << (1 + UR(option::GetHavocStackPow2(state), state.rand_fd));
 
-        /* out_buf might have been mangled a bit, so let's restore it to its
-           original size and shape. */
+    state.stage_cur_val = use_stacking;
+    mutator.Havoc(use_stacking, state.extras, state.a_extras, mutop_optimizer,
+                  custom_cases);
 
-        mutator.RestoreHavoc();
+    u64 havoc_finds = state.queued_paths + state.unique_crashes;
 
-        /* If we're finding new stuff, let's run for a bit longer, limits
-           permitting. */
+    if (this->CallSuccessors(mutator.GetBuf(), mutator.GetLen())) return true;
 
-        if (state.queued_paths != havoc_queued) {
-            if (perf_score <= option::GetHavocMaxMult(state) * 100) {
-                state.stage_max *= 2;
-                perf_score *= 2;
-            }
+    havoc_finds = state.queued_paths + state.unique_crashes - havoc_finds;
```

Comment:
I don't like to see the same variable used different ways. 
I prefer something like, `u64 prev_havoc_finds = state.queued_paths + state.unique_crashes;` and `havoc_finds_diff = state.queued_paths + state.unique_crashes - prev_havoc_finds;`.

Reply:
fixed in 3aed518b55a335259bd324e710f4d0bac862ddef